src/legacy/server/capabilities/merge_capabilities.ts

```diff
@@ -17,23 +17,15 @@
  * under the License.
  */
 
+import { merge } from 'lodash';
 import { Capabilities } from '../../../core/public';
 
 export const mergeCapabilities = (...sources: Array<Partial<Capabilities>>): Capabilities =>
-  sources.reduce(
-    (capabilities: Capabilities, source) => {
-      Object.entries(source).forEach(([key, value = {}]) => {
-        capabilities[key] = {
-          ...value,
-          ...capabilities[key],
-        };
-      });
-
-      return capabilities;
-    },
+  merge(
```

Comment:
The previous implementation wasn't preventing duplicate definitions of the same first-level ui capability: https://codepen.io/kobelb/pen/MMWROw?editors=0012. We take advantage of this for building the `catalogue` ui capabilities: some of them are hard-coded [here](https://github.com/elastic/kibana/blob/590862ec635fbbbdde6a188c3fe1eabf6b12a3be/src/legacy/core_plugins/kibana/index.js#L261-L268) and we add to this list within the xpack_main plugin [here](https://github.com/elastic/kibana/blob/590862ec635fbbbdde6a188c3fe1eabf6b12a3be/x-pack/plugins/xpack_main/server/lib/ui_capabilities_for_features.ts#L32-L41).

> While it's true that merge won't allow a plugin to delete capabilities defined by other plugins, the addition of capabilities to other plugin's "buckets" is likely not something we want to encourage/support.

I agree, but I don't believe those restrictions where previously enforced, at least not when using `injectDefaultVars` (when we first implemented FC) or the new `uiCapabilities`. We have some enforcement on this within `xpack_main` [here](https://github.com/elastic/kibana/blob/master/x-pack/plugins/xpack_main/server/lib/ui_capabilities_for_features.ts), but this is only for merging the ui capabilities that we derive from the feature registration.

I can look at changing spaces to not use `uiCapabilities` to define the management section, and instead augment [uiCapabilitiesForFeatures](https://github.com/elastic/kibana/blob/master/x-pack/plugins/xpack_main/server/lib/ui_capabilities_for_features.ts), but we'll still need the ability to merge all of these "deeply".

> Since management is a defined key of the Capabilities interface, how do you feel about allowing the merge for this key, but not for the rest?

I think the proper solution to this problem is to follow through on https://github.com/elastic/kibana/pull/36221 and make it so that each of these "ui capability sections" is able to define it's own rules and so we can enforce these restrictions throughout the system, I was trying to prevent making this a dependency for fixing this specific issue though.

I added tests with 4d21c3436fc0dbfdea5f130e881f5ede1369f828 to ensure we aren't clobbering anything, and added additional safeguards. Does this assuage any of your previous concerns?

Reply:
> I can look at changing spaces to not use uiCapabilities to define the management section, and instead augment uiCapabilitiesForFeatures, but we'll still need the ability to merge all of these "deeply".

So... we can't do this at the moment because we don't register a "Spaces" feature, and these actions are hard-coded here: https://github.com/elastic/kibana/pull/38472/files#diff-f5ed4149300a16c9d74494d7b470b450R68. I don't know what I was thinking previously.

Reply:
> I think the proper solution to this problem is to follow through on #36221 and make it so that each of these "ui capability sections" is able to define it's own rules and so we can enforce these restrictions throughout the system, I was trying to prevent making this a dependency for fixing this specific issue though.

Agreed, thanks for the thoughtful followup!

> I added tests with 4d21c34 to ensure we aren't clobbering anything, and added additional safeguards. Does this assuage any of your previous concerns?

❤️